### PR TITLE
[refactor] Statically initialize Token.tochars

### PR DESCRIPTION
**Pull request by yebblies**
Bonus points to anyone who can pick out the two duplicate inits in the original list!

### CODE REVIEW

File: src/tokens.d [hunk 1]
```diff
@@ -577,7 +577,112 @@ extern (C++) struct Token
...
+    static __gshared const(char)*[TOKMAX] tochars =
+    [
+        TOKeof: "EOF",
+        TOKlcurly: "{",
+        TOKrcurly: "}",
+        TOKlparen: "(",
+        TOKrparen: ")",
+        TOKlbracket: "[",
+        TOKrbracket: "]",
+        TOKsemicolon: ",",
```
Comment (Geod24) [inline src/tokens.d]: Oups.
Also, bonus point on "!is" ? ;)

Comment (yebblies) [inline src/tokens.d]: Fixed!

**Comment by ZombineDev** [comment]
@yebblies how about:
```
enum Token : string
{
    eof = "EOF",
    lcurly = "{",
    rcurly = "}",
    // and so on...
}
```
Bonus: you can use final switch after you are sure that input is valid.

**Comment by ZombineDev** [comment]
@yebblies and for compatability with existing code you can add :
mixin (generateEnumAliases!(Token, "TOK"));

**Comment by yebblies** [comment]
That would still be slower than switching on integers, no?

**Comment by ZombineDev** [comment]
So the disadvantage is performance, beacause currently it's not lowered in an opimal way?

Off topic, but I'm wondering if the following enhancement would be accepted:
1. Compiler puts all enum member values in a lexicograpically sorted table (or some perfect hashing scheme is used) in the read-only section of the object file.
2. A hidden "__enum_out_of_range" member is added by the compiler.
2. Each enum memver will have a corresponding integer. (The index in the table).
  + Instances of such enum will be represented as integers in memory.
  + E.g. `Token x = Token.eof; assert (cast(ulong)x == 0);`, or some other integer, the values are an implementation detail and should not be directly relied upon.
3. On assignment by value of the semantically underlying enum type (e.g. `static assert (is(EnumType!Token == string)); Token a = "eof";`  a lookup in made in the corresponding table.
  + If the value is found `x` will be assigned the index of enum member in the table.
  + Otherwise `x` will be assigned the value of the hidden compiler-generated member `__enum_out_of_range`.

4.`final switch (x)` continues to work as it has with integers:
  + If `x == Token.__enum_out_of_range`, `SwitchError` is thrown.
  + Otherwise (using integer switch table) the instruction pointern jumps to the corresponding case statement.
5. `switch (x)` work in a similar way:
  + The only difference is that if `x == Token.__enum_out_of_range` the `default` case is executed.
6. Compiler disallows initialization or assingment of values that are not the same type as the semantically underlying enum type.
7. I know you had similar proposal for built-in CT AAs, but I believe my proposal is better, because can help AA as a library and doesn't add baggage to the existing implementation.
8. Destroy!

**Comment by DmitryOlshansky** [comment]
> Destroy!

Or just use perfect hashing for strings.

**Comment by yebblies** [comment]
Just a gentle reminder that none of this has anything to do with this PR, which I would like merged.

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
Thanks!

**Comment by yebblies** [comment]
@ZombineDev That proposal sounds useful, but it doesn't look like it couldn't be done entirely in the library.  My guess is that it would be rejected as a language enhancement because of this.  Tables like this are not generally difficult to maintain, and when they are they can usually be simplified with mixins.

**Pull request merged**